src/HexagonOptimize.cpp

```diff
@@ -0,0 +1,1084 @@
+        if (is_native_interleave(x)) {
+            return x.as<Call>()->args[0];
+        } else if (x.type().is_scalar() || x.as<Broadcast>()) {
+            return x;
+        }
+        const Variable *var = x.as<Variable>();
+        if (var) {
+            internal_assert(vars.contains(var->name + ".deinterleaved"));
+            return Variable::make(var->type, var->name + ".deinterleaved");
+        }
+        internal_error << "Expression '" << x << "' does not yield an interleave.\n";
+        return x;
+    }
+
+    template <typename T>
+    void visit_binary(const T* op) {
+        Expr a = mutate(op->a);
+        Expr b = mutate(op->b);
+        // We only want to pull out an interleave if at least one of
+        // the operands is an actual interleave. Furthermore, we can
+        // only attempt to do this if we are allowing the expr to be
+        // interleaved via interleave_expr, or the result is not boolean.
+        bool can_interleave = op->type.bits() != 1;
+        if ((can_interleave || allow_interleave_expr == 0) && yields_removable_interleave({a, b})) {
+            a = remove_interleave(a);
+            b = remove_interleave(b);
+            expr = T::make(a, b);
+            if (can_interleave) {
+                expr = native_interleave(expr);
+            } else {
+                internal_assert(!interleave_expr);
+                interleave_expr = true;
+            }
+        } else if (!a.same_as(op->a) || !b.same_as(op->b)) {
+            expr = T::make(a, b);
+        } else {
+            expr = op;
+        }
+    }
+
+    void visit(const Add *op) { visit_binary(op); }
+    void visit(const Sub *op) { visit_binary(op); }
+    void visit(const Mul *op) { visit_binary(op); }
+    void visit(const Div *op) { visit_binary(op); }
+    void visit(const Mod *op) { visit_binary(op); }
+    void visit(const Min *op) { visit_binary(op); }
+    void visit(const Max *op) { visit_binary(op); }
+    void visit(const EQ *op) { visit_binary(op); }
+    void visit(const NE *op) { visit_binary(op); }
+    void visit(const LT *op) { visit_binary(op); }
+    void visit(const LE *op) { visit_binary(op); }
+    void visit(const GT *op) { visit_binary(op); }
+    void visit(const GE *op) { visit_binary(op); }
+
+    // These next 3 nodes should not exist if we're vectorized, they
+    // should have been replaced with bitwise operations.
+    void visit(const And *op) {
+        internal_assert(op->type.is_scalar());
+        IRMutator::visit(op);
+    }
+    void visit(const Or *op) {
+        internal_assert(op->type.is_scalar());
+        IRMutator::visit(op);
+    }
+    void visit(const Not *op) {
+        internal_assert(op->type.is_scalar());
+        IRMutator::visit(op);
+    }
+
+    void visit(const Select *op) {
+        Expr true_value = mutate(op->true_value);
+        Expr false_value = mutate(op->false_value);
+
+        Expr cond;
+        if (yields_removable_interleave({true_value, false_value})) {
```

Comment:
Also, there's a minor performance bug here: yields_removable_interleave checks to see if all of its operands yield an interleave, but also that one of them *is* an interleave. However, we still want to remove the interleave from the condition as long as the true/false values both yield an interleave, even if neither of them actually is an interleave itself.